base/missing.jl

```diff
@@ -189,6 +189,99 @@ function Base.iterate(itr::SkipMissing, state...)
     item, state
 end
 
+# Optimized mapreduce implementation
+mapreduce(f, op, itr::SkipMissing{<:AbstractArray}) = _mapreduce(f, op, IndexStyle(itr.x), itr)
+
+"Sentinel indicating that no non-missing value was encountered."
+struct AllMissing end
+
+function _mapreduce(f, op, ::IndexLinear, itr::SkipMissing{<:AbstractArray})
+    A = itr.x
+    local ai
+    inds = LinearIndices(A)
+    i = first(inds)
+    ilast = last(inds)
+    while i <= ilast
+        @inbounds ai = A[i]
+        ai === missing || break
+        i += 1
+    end
+    i > ilast && return mapreduce_empty(f, op, eltype(itr))
+    a1 = ai
+    i += 1
+    while i <= ilast
+        @inbounds ai = A[i]
+        ai === missing || break
+        i += 1
+    end
+    i > ilast && return mapreduce_first(f, op, a1)
+    # We know A contains at least two non-missing entries, therefore AllMissing() is not
+    # a possible return value: the check provides that information to inference
+    s = mapreduce_impl(f, op, itr, first(inds), last(inds))
+    s isa AllMissing && error("got AllMissing for an array with non-missing values")
+    return s
+end
+
+_mapreduce(f, op, ::IndexCartesian, itr::SkipMissing) = mapfoldl(f, op, itr)
+
+mapreduce_impl(f, op, A::SkipMissing, ifirst::Integer, ilast::Integer) =
+    mapreduce_impl(f, op, A, ifirst, ilast, pairwise_blocksize(f, op))
+
+@noinline function mapreduce_impl(f, op, itr::SkipMissing{<:AbstractArray},
+                                  ifirst::Integer, ilast::Integer, blksize::Int)
+    A = itr.x
+    if ifirst == ilast
+        @inbounds a1 = A[ifirst]
+        if a1 === missing
+            return AllMissing()
+        else
+            return mapreduce_first(f, op, a1)
+        end
+    elseif ifirst + blksize > ilast
+        # sequential portion
+        local ai
+        i = ifirst
+        while i <= ilast
+            @inbounds ai = A[i]
+            ai === missing || break
+            i += 1
+        end
+        i > ilast && return AllMissing()
+        a1 = ai::eltype(itr)
+        i += 1
+        while i <= ilast
+            @inbounds ai = A[i]
+            ai === missing || break
+            i += 1
+        end
+        i > ilast && return mapreduce_first(f, op, a1)
+        a2 = ai::eltype(itr)
+        # Unexpectedly, the following assertion allows SIMD instructions to be emitted
+        A[i]::eltype(itr)
```

Comment:
The fact that this line is required to get SIMD is really intriguing. In the present case, we know `A[i]` is non-missing, so it should be perfectly fine to keep. And this assertion doesn't seem to force the compiler to incorrectly assume that all entries in `A` are non-missing, since we do get `missing` values in the loop below.

Minimal reproducer:
```julia
function g(A)
    v = 0
    # Uncomment to get SIMD
    # A[1]::Base.nonmissingtype(eltype(A))
    @simd for i = 1:length(A)
        @inbounds ai = A[i]
        if ai !== missing
            v += ai
        end
    end
    return v
end

@code_llvm g([1,missing])
```

Reply:
Maybe open the minimal reproducer in a standalone issue in order to get more attention?

Reply:
Yes, I can file a separate issue if that's useful, but @vtjnash can probably write a much more useful description than I can.